src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -853,61 +831,90 @@ protected SnowflakeStreamingIngestChannel getChannel() {
+   * records from Kafka and once threshold has reached, we would call insertRows API to insert into
+   * Snowflake.
    *
-   * <p>We would remove this long lived buffer logic in separate commit. SNOW-473896
+   * <p>We would transform kafka records to Snowflake understood records (In JSON format) just
+   * before calling insertRows API.
    */
   @VisibleForTesting
   protected class StreamingBuffer extends PartitionBuffer<List<Map<String, Object>>> {
-    // Used to buffer rows per channel
-    // Map has key of column name and Object is its value
-    // For KC, it will be metadata and content columns.
-    // Every List element corresponds to one record(offset) of a partition
-    private final List<Map<String, Object>> tableRows;
-
+    // Records coming from Kafka
     private final List<SinkRecord> sinkRecords;
 
     StreamingBuffer() {
       super();
-      tableRows = new ArrayList<>();
       sinkRecords = new ArrayList<>();
     }
 
     @Override
-    public void insert(SinkRecord record) {
-
-      // has two keys(two columns), each of the values are JsonNode
-      Map<String, Object> tableRow = recordService.getProcessedRecordForStreamingIngest(record);
+    public void insert(SinkRecord kafkaSinkRecord) {
 
-      if (tableRows.size() == 0) {
-        setFirstOffset(record.kafkaOffset());
+      if (sinkRecords.isEmpty()) {
+        setFirstOffset(kafkaSinkRecord.kafkaOffset());
       }
-
-      tableRows.add(tableRow);
-      sinkRecords.add(record);
+      sinkRecords.add(kafkaSinkRecord);
 
       // probably do below things in a separate method.
       // call it collect buffer metrics
       setNumOfRecords(getNumOfRecords() + 1);
-      setLastOffset(record.kafkaOffset());
+      setLastOffset(kafkaSinkRecord.kafkaOffset());
       // need to loop through the map and get the object node
-      tableRow.forEach(
-          (key, value) -> {
-            setBufferSizeBytes(
-                getBufferSizeBytes() + ((String) value).length() * 2L); // 1 char = 2 bytes
-          });
+      // update size of buffer
+      setBufferSizeBytes(
+          getBufferSizeBytes() + StreamingUtils.getApproxSizeOfRecordInBytes(kafkaSinkRecord));
     }
 
-    /* Get all rows which are present in this list. Each map corresponds to one row whose keys are column names. */
+    /**
+     * Get all rows which which were buffered into this buffer. Each map corresponds to one row
+     * whose keys are column names and values are corresponding data in that column.
+     *
+     * <p>This goes over through all buffered kafka records and transforms into JsonSchema and
+     * JsonNode Check {@link #handleNativeRecord(SinkRecord, boolean)}
+     *
+     * @return List of Map of String(COLUMN_NAME) and Object
+     */
     @Override
     public List<Map<String, Object>> getData() {
+      final List<Map<String, Object>> snowflakeTableRowsFromSinkRecords = new ArrayList<>();
+      for (SinkRecord kafkaSinkRecord : sinkRecords) {
```

Comment:
could we do this in a parallel stream to improve throughput?

Reply:
Yeah sounds like we can use multiple threads here. I dont like parallel streams because it hides away which thread did it and is kind of a blackbox. 
https://stackoverflow.com/questions/20375176/should-i-always-use-a-parallel-stream-when-possible

Although this does qualify for parallelism. 

Would it be okay to send out of order offsets to insertRows? as far as last offset stays the same?

Reply:
I confirmed the code inside this is thread safe to begin with. Also, parallelstream uses 1 less as number of threads from forkjoin pool. So let me just use it..

Reply:
No, i think we want to keep the ordering, you will probably need an array and put the processed row to the original index.

Reply:
Let me do it as part of separate PR actually as this is just perf improvement which we cant justify without testing. (In theory yes, it will be better but I want to do testing too) [SNOW-555624](https://snowflakecomputing.atlassian.net/browse/SNOW-555624)